### ISSUE TITLE
Type inference defaulting to dynamic can cause hard to find issues

### ISSUE THREAD

**Issue by andreashaese** [dart-lang/language#150]
This is my first post here, apologies if I'm not correctly following protocol. I've run into an issue that is caused by Dart's silent defaulting to `dynamic` if type inference doesn't find a concrete type. Consider the following code:

    Type typeOf<T>() => T;
    
    class Value<T> extends InheritedWidget {
      final T value;
      
      Value({this.value, Widget child}) : super(child: child);
      
      @override bool updateShouldNotify(_) => true;
      
      static T of<T>(BuildContext context) {
        final type = typeOf<Value<T>>();
        final provider = context.inheritFromWidgetOfExactType(type) as Value<T>;
        return provider.value; // <-- (1)
      }
    }
    
    class StatefulValue<T> extends StatefulWidget {
      final T Function(BuildContext) valueBuilder;
      final Widget child;
      
      StatefulValue({this.valueBuilder, this.child}) : super();
      
      @override createState() => _StatefulValueState(); // <-- (2)
    }
    
    class _StatefulValueState<T> extends State<StatefulValue<T>> {
      T _value;
      
      @override initState() {
        super.initState();
        _value = widget.valueBuilder(context);
      }
      
      @override build(context) => Value<T>(value: _value, child: widget.child);
    }

This is then used like so:

    class HomePage extends StatelessWidget {
      @override build(context) => StatefulValue<String>(
        valueBuilder: (_) => 'Hello!',
        child: Builder(
          builder: (context) => Scaffold(
            appBar: AppBar(
              title: Text(Value.of<String>(context)),
            ),
          ),
        ),
      );
    }

The details of the example don't matter, it just aims to demonstrate a subtle bug that's hidden within typical "Flutter noise": The code compiles without warnings, but instead of showing the expected 'Hello!' Text in the app bar, a runtime error is thrown: `"The getter 'value' was called on null."` (see `(1)`).

It took me quite a while to find the fix (see `(2)`):

    @override createState() => _StatefulValueState<T>();

In hindsight it makes perfect sense: without the `<T>` annotation, Dart defaults to `_StatefulValueState<dynamic>`. An easily made mistake that creates a cascade of issues that's very hard to backtrack. In fact, now that I'm aware of this, I'm not at all sure I haven't made similar mistakes elsewhere that just haven't surfaced as bugs yet.

Testing surely helps, but I wonder if the Dart languge itself could be more helpful in such cases: If the analyzer fails to infer a concrete type, that's probably because I did something wrong unintentionally, and I surely would benefit from knowing about it. I wonder, would a compiler warning (or even a compile-time error) be feasible?

**Comment by zoechi** [comment]
Might be #31410
https://www.dartlang.org/guides/language/analysis-options#enabling-additional-type-checks

**Comment by andreashaese** [comment]
Oh perfect, thank you! In my opinion it's the most sensible default, but a quite hidden feature, so maybe there's value in making this opt-out instead of opt-in?

**Comment by munificent** [comment]
Yes, this is something we've wanted to push on for a while but haven't had the time. When we first introduced the new type system ("strong mode" at the time), we had to balance catching more errors statically with making it not too painful to migrate existing Dart 1 code to the new type system.

One of the places where we were lax is silently inserting dynamic in places where type inference couldn't come up with a better type. At the time, there was a lot of code that would have been broken if we'd made that an error, and we wanted to minimize the migration. Now that most users are on the new type system and want even better static checking, I think it's worth trying to tighten things a bit more.

There's some work going on around adding a new set of strictness checks that users can opt into and this would be covered by one of them. We can't make this the default because that would break existing code. But, over time, if enough users opt in, we could consider making it the default in a later breaking change release of the language.

That's how strong mode was shipped: You could opt in during several 1.x releases and then 2.0 made it required.

**Comment by andreashaese** [comment]
Of course, I understand. Having used Swift productively since v1.0, I know how painful, and expensive, breaking language changes can be.

However, since a couple of versions Swift has been striking what I find a good balance between improving (but breaking) changes, and keeping existing code compatible through robust automatic and/or interactive migration assistants. E.g. in the current case, a naïve migrator could suggest to fill in `dynamic` where the compiler sees it missing. But I’m not telling you something new here.

At the end of the day, I think it can’t be overstated how much superior compile-time errors are compared to runtime errors in terms of development efficiency, not to mention developer and end user experience. I’m very pleased by the steps Dart has already taken in this direction, and I’m looking forward to what more is to come!

**Comment by kevlar700** [comment]
This issue is close enough that perhaps it is better than the mailing list for my comment or creating a separate issue. The answer should resolve this too.

It just so happened that I turned off explicit-dynamic for half a day and at the same time decided to see if function arguments could be shortened like in the go language and it seemed to work.

(first, second, third string, third, fourth int)

When I decide to turn explicit-dynamic back on. I suddenly realised that the equivalents of first, second and third were now dynamics!

**Comment by eernstg** [comment]
That depends on the context. If there is no information about the intended parameter type of a function, it will be `dynamic`. But the information could come from several different sources: A function literal could be assigned to a variable or passed as an actual argument, and the declared type of the variable/parameter is then the context for the function literal. A method could override another method, and it would then "inherit" the missing parts of the signature:

```dart
void printStaticType<X>(List<X> xs) { print(X); }

class A {
  void m(A first, B second, String third, bool fourth, int fifth) {}
}

class B extends A {
  m(first, second, third, fourth, fifth) {
    printStaticType([first]);
    printStaticType([second]);
    printStaticType([third]);
    printStaticType([fourth]);
    printStaticType([fifth]);
  }
}

main() {
  B().m(A(), B(), '', true, 5);

  void Function(A, B, String, bool, int) f = (_1, _2, _3, _4, _5) {};
  print(f.runtimeType);
}
```
